components/brave_shields/browser/ad_block_service.cc

```diff
@@ -337,41 +245,70 @@ AdBlockService::AdBlockService(
+    const base::FilePath& profile_dir)
     : local_state_(local_state),
       locale_(locale),
+      profile_dir_(profile_dir),
+      subscription_download_manager_getter_(
+          std::move(subscription_download_manager_getter)),
       component_update_service_(cus),
       task_runner_(task_runner),
-      custom_filters_service_(nullptr, base::OnTaskRunnerDeleter(task_runner_)),
-      default_service_(nullptr, base::OnTaskRunnerDeleter(task_runner_)),
-      subscription_service_manager_(std::move(subscription_service_manager)) {
+      default_engine_(std::unique_ptr<AdBlockEngine, base::OnTaskRunnerDeleter>(
+          new AdBlockEngine(),
+          base::OnTaskRunnerDeleter(GetTaskRunner()))),
+      additional_filters_engine_(
+          std::unique_ptr<AdBlockEngine, base::OnTaskRunnerDeleter>(
+              new AdBlockEngine(),
+              base::OnTaskRunnerDeleter(GetTaskRunner()))) {
   // Initializes adblock-rust's domain resolution implementation
   adblock::SetDomainResolver(AdBlockServiceDomainResolver);
 
-  default_filters_provider_ =
-      std::make_unique<brave_shields::AdBlockComponentFiltersProvider>(
-          component_update_service_, g_ad_block_component_id_,
-          g_ad_block_component_base64_public_key_, kAdBlockComponentName);
+  additional_filters_combinator_ =
+      std::make_unique<AdBlockCombinatorFiltersProvider>();
+
   resource_provider_ =
       std::make_unique<brave_shields::AdBlockDefaultResourceProvider>(
           component_update_service_);
   filter_list_catalog_provider_ =
       std::make_unique<brave_shields::AdBlockFilterListCatalogProvider>(
           component_update_service_);
+
+  default_filters_provider_ =
+      std::make_unique<brave_shields::AdBlockComponentFiltersProvider>(
+          component_update_service_, g_ad_block_component_id_,
+          g_ad_block_component_base64_public_key_, kAdBlockComponentName);
+  regional_service_manager_ =
+      brave_shields::AdBlockRegionalServiceManagerFactory(
+          additional_filters_combinator_.get(), local_state_, locale_,
+          component_update_service_, GetTaskRunner());
+  subscription_service_manager_ =
+      std::make_unique<brave_shields::AdBlockSubscriptionServiceManager>(
+          additional_filters_combinator_.get(), local_state_, GetTaskRunner(),
+          std::move(subscription_download_manager_getter_), profile_dir_);
   custom_filters_provider_ =
       std::make_unique<brave_shields::AdBlockCustomFiltersProvider>(
           local_state_);
+  additional_filters_combinator_->CombinateProvider(
+      custom_filters_provider_.get());
+
+  default_service_observer_ = std::make_unique<SourceProviderObserver>(
+      default_engine_->AsWeakPtr(), default_filters_provider_.get(),
+      resource_provider_.get(), GetTaskRunner());
+  additional_filters_service_observer_ =
+      std::make_unique<SourceProviderObserver>(
+          additional_filters_engine_->AsWeakPtr(),
```

Comment:
why does this need to be WeakPtr? `additional_filters_service_observer_` will be destroyed before `additional_filters_engine_`

Reply:
It was already that way before, when there were engines all over the place. But it makes sense to change it now.